x-pack/plugins/endpoint/common/types.ts

```diff
@@ -13,6 +13,8 @@ import { alertingIndexGetQuerySchema } from './schema/alert_index';
  */
 export type Immutable<T> = T extends undefined | null | boolean | string | number
   ? T
+  : unknown extends T
+  ? unknown
```

Comment:
this is safe because anything union'ed w/ `unknown` is `unknown`
![image](https://user-images.githubusercontent.com/35559/79164128-6b86cf80-7dae-11ea-90da-b3158a52ce3a.png)

Reply:
why do you need to do `: unknown extends T ? unknown`

Reply:
`Immutable` is meant to be able to accept anything. If you have a type such as:
```
type nerds = {
  a: string,
  b: unknown
}
```
Then when we recurse to `b`, the value will be `Immutable<unknown>`. Without special handling this will fallback to `ImmutableObject<unknown>`. `ImmutableObject<unknown>` doesn't work (maybe `ImmutableObject` should have stricter type parameter constraints.) 

> ["Anything is assignable to `unknown`, but `unknown` isn’t assignable to anything but itself and `any` without a type assertion or a control flow based narrowing."](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#new-unknown-top-type)

Therefore `T extends unknown` wouldn't be a good way to handle `unknown` as it will always be true. e.g.:

`type itsTrue = 1 extends unknown ? true : false`

If you have any union with `unknown`, it is the same as unknown. This is because that union can accept anything (everything is assignable to unknown.) So if we want to have `Immutable<unknown>` alias `unknown`, we can check if `unknown` can be assigned to `T`.